applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.h

```diff
@@ -70,32 +92,188 @@ class KRATOS_API(FLUID_DYNAMICS_APPLICATION) DistanceSmoothingProcess : public P
+
+        CreateSolutionStrategy(p_builder_solver);
+    }
+
+    /// Constructor.
+    DistanceSmoothingProcess(
+        ModelPart& rModelPart,
+        typename TLinearSolver::Pointer p_linear_solver,
+        BuilderSolverPointerType pBuilderAndSolver,
+        const bool AllConditionsAsBoundary = true)
+        : Process(),
+        mrModelPart(rModelPart),
+        mrModel(rModelPart.GetModel()),
+        mAuxModelPartName("smoothing_model_part"),
+        mAllConditionsAsBoundary(AllConditionsAsBoundary),
+        mAuxModelPartIsInitialized(false)
+    {
+        // Generate an auxilary model part and populate it by elements of type DistanceSmoothingElement
+        CreateAuxModelPart();
+
+        CreateSolutionStrategy(pBuilderAndSolver);
+    }
 
     /// Constructor with Kratos parameters.
     DistanceSmoothingProcess(
         ModelPart& rModelPart,
-        Parameters Parameters);
+        Parameters Parameters)
+        : DistanceSmoothingProcess(
+        rModelPart,
+        LinearSolverFactory<TSparseSpace, TDenseSpace>().Create(Parameters["linear_solver_settings"])
+    ){}
 
     /// Constructor with Kratos model
     DistanceSmoothingProcess(
         Model& rModel,
-        Parameters Parameters);
+        Parameters Parameters)
+        : DistanceSmoothingProcess(
+        rModel.GetModelPart(Parameters["model_part_name"].GetString()),
+        Parameters
+    ){}
 
     /// Destructor.
     ~DistanceSmoothingProcess() override
     {
-        if(mrModel.HasModelPart( mAuxModelPartName ))
-            mrModel.DeleteModelPart( mAuxModelPartName );
+        Clear();
     }
 
     ///@}
     ///@name Operators
     ///@{
 
-    void Execute() override;
+    void Execute() override
+    {
+        KRATOS_TRY;
+
+        if(mAuxModelPartIsInitialized == false){
+            CreateAuxModelPart();
+        }
+
+        auto& r_smoothing_model_part = mrModel.GetModelPart( mAuxModelPartName );
+
+        block_for_each(r_smoothing_model_part.Nodes(), [&](Node<3>& rNode){
+                rNode.Free(DISTANCE);
+                const double distance = rNode.FastGetSolutionStepValue(DISTANCE);
+                rNode.FastGetSolutionStepValue(DISTANCE, 1) = distance;
+            });
+
+        mp_solving_strategy->Solve();
+
+        auto &r_communicator = r_smoothing_model_part.GetCommunicator();
+        r_communicator.SynchronizeCurrentDataToMin(DISTANCE);
+
+        block_for_each(r_smoothing_model_part.Nodes(), [&](Node<3>& rNode){
+                rNode.SetValue( DISTANCE, rNode.FastGetSolutionStepValue(DISTANCE)
+                    - rNode.FastGetSolutionStepValue(DISTANCE, 1) ); // Corrected distance difference
+            });
+
+        auto& r_data_comm = r_smoothing_model_part.GetCommunicator().GetDataCommunicator();
+        GlobalPointersVector< Node<3 > > gp_list;
+
+        for (int i_node = 0; i_node < static_cast<int>(r_smoothing_model_part.NumberOfNodes()); ++i_node){
```

Comment:
```suggestion
        const std::size_t n_nodes = r_smoothing_model_part.NumberOfNodes();
        for (std::size_t i_node = 0; i_node < n_nodes; ++i_node){
```
To avoid the cast each time the exit condition is evaluated.

Reply:
Done.